Source/Core/Core/Slippi/SlippiPlayback.cpp

```diff
@@ -0,0 +1,272 @@
+#include <memory>
+#include <mutex>
+
+#ifdef _WIN32
+#include <share.h>
+#endif
+
+#include "Common/Logging/Log.h"
+#include "Core/Core.h"
+#include "Core/HW/EXI_DeviceSlippi.h"
+#include "Core/NetPlayClient.h"
+#include "Core/State.h"
+#include "SlippiPlayback.h"
+
+#define FRAME_INTERVAL 900
+#define SLEEP_TIME_MS 8
+
+std::unique_ptr<SlippiPlaybackStatus> g_playbackStatus;
+extern std::unique_ptr<SlippiReplayComm> g_replayComm;
+
+static std::mutex mtx;
+static std::mutex seekMtx;
+static std::mutex diffMtx;
+static std::unique_lock<std::mutex> processingLock(diffMtx);
+static std::condition_variable condVar;
+static std::condition_variable cv_waitingForTargetFrame;
+static std::condition_variable cv_processingDiff;
+static std::atomic<int> numDiffsProcessing(0);
+
+s32 emod(s32 a, s32 b)
+{
+	assert(b != 0);
+	int r = a % b;
+	return r >= 0 ? r : r + std::abs(b);
+}
+
+std::string processDiff(std::vector<u8> iState, std::vector<u8> cState)
+{
+	INFO_LOG(SLIPPI, "Processing diff");
+	numDiffsProcessing += 1;
+	cv_processingDiff.notify_one();
+	std::string diff = std::string();
+	open_vcdiff::VCDiffEncoder encoder((char *)iState.data(), iState.size());
+	encoder.Encode((char *)cState.data(), cState.size(), &diff);
+
+	INFO_LOG(SLIPPI, "done processing");
+	numDiffsProcessing -= 1;
+	cv_processingDiff.notify_one();
+	return diff;
+}
+
+SlippiPlaybackStatus::SlippiPlaybackStatus()
+{
+	shouldJumpBack = false;
+	shouldJumpForward = false;
+	inSlippiPlayback = false;
+	shouldRunThreads = false;
+	isHardFFW = false;
+	isSoftFFW = false;
+	lastFFWFrame = INT_MIN;
+	currentPlaybackFrame = INT_MIN;
+	targetFrameNum = INT_MAX;
+	latestFrame = Slippi::GAME_FIRST_FRAME;
+}
+
+void SlippiPlaybackStatus::startThreads()
+{
+	shouldRunThreads = true;
+	m_savestateThread = std::thread(&SlippiPlaybackStatus::SavestateThread, this);
+	m_seekThread = std::thread(&SlippiPlaybackStatus::SeekThread, this);
+}
+
+void SlippiPlaybackStatus::prepareSlippiPlayback()
+{
+	// block if there's too many diffs being processed
+	while (numDiffsProcessing > 3)
+	{
+		INFO_LOG(SLIPPI, "Processing too many diffs, blocking main process");
+		cv_processingDiff.wait(processingLock);
+	}
+
+	// Unblock thread to save a state every interval
+	if ((currentPlaybackFrame + 122) % FRAME_INTERVAL == 0)
+		condVar.notify_one();
+
+	// TODO: figure out why sometimes playback frame increments past targetFrameNum
+	if (inSlippiPlayback && currentPlaybackFrame >= targetFrameNum)
```

Comment:
This is the line that fixes like half of everything. 

I'm pretty sure this is a race condition between save state loading and slippi exi DMAWrite. sometimes DMAWrite will overwrite currentPlaybackFrame and increment it even though the plan is to rewind the state, changing currentPlaybackFrame and then fast forward to the proper frame.